// magic-reverse-proxy/index.js
const ProxyServer = require('./libs/proxysrv');
const Express = require('express');
const Https = require('https');
const FileSystem = require('fs');
const Wildcard = require('wildcard');

const stdcfg = {
  enable_hsts: false,
  allow_unknown_host: true,
  http: {
    port: 80,
    enabled: true,
    start_callback: function(){},
    middlewares: []
  },
  https: {
    port: 443,
    enabled: false,
    start_callback: function(){},
    middlewares: [],
    sslkey: '',
    sslcert: '',
    sslchain: ''
  },
  proxies: [],
  default_proxy: {
    destination: [],
    timeout: 10000,
    round: 0
  }
};

function is_set(val){
  return !((typeof val == 'undefined') || (val === null));
}

/**Generate default options for the input specified.
 * 
 * @param {stdcfg} options Options of proxy.
 */
function default_options(options) {
  let opts                = options || stdcfg;
  opts.enable_hsts        = !is_set(opts.enable_hsts) ? false : opts.enable_hsts;
  opts.allow_unknown_host = !is_set(opts.allow_unknown_host) ? true : opts.allow_unknown_host;
  opts.http               = opts.http || stdcfg.http;
  opts.https              = opts.https || stdcfg.https;
  opts.proxies            = opts.proxies || [];
  opts.default_proxy      = opts.default_proxy || stdcfg.default_proxy;
  return opts;
}

/**Check if the remote host used to connect to the server is allowed.
 * If not allowed, then the TCP connection will be closed with RST.
 * 
 * @param {stdcfg} defOpts Options of proxy.
 */
function check_host(defOpts){
  return (req,res,next) => {
    let is_known_host = false;
    defOpts.proxies.forEach((proxyObj, idx, proxies) => {
      if (!is_known_host && Wildcard(String(proxyObj.domain).toUpperCase(), String(req.hostname).toUpperCase())) {
        is_known_host = true;
      }
    });
    if (!is_known_host){ try{ res.connection.destroy(); }catch(e){} } else { return next(); }
  }
}

/**Create a new proxy object that contains the express app, appssl and the bind() function that starts all servers and listen on local port.
 * 
 * @param {stdcfg} options Options of proxy.
 */
function createProxy(options){
  let defOptions = default_options(options);
  let proxy = {
    app: Express(),
    appssl: Express(),
    config: defOptions,
    /**Bind the proxy on local port spcified in options of createProxy() function.
     * @returns {void} 
     */
    bind: function(){
      if (defOptions.http.enabled == true) {
        this.app.listen(defOptions.http.port, defOptions.http.start_callback);
      }
      if (defOptions.https.enabled == true) {
        Https.createServer({
          key: FileSystem.readFileSync(defOptions.https.sslkey),
          cert: FileSystem.readFileSync(defOptions.https.sslcert),
          ca: FileSystem.readFileSync(defOptions.https.sslchain)
        }, this.appssl).listen(defOptions.https.port, defOptions.https.start_callback);
      }
      defOptions.http.middlewares.forEach((middleware, idx, arr) => {
        this.app.use(middleware);
      });
      defOptions.https.middlewares.forEach((middleware, idx, arr) => {
        this.appssl.use(middleware);
      });
      if (this.config.http.enabled && !this.config.allow_unknown_host){
        this.app.use(check_host(defOptions));
      }
      if (this.config.https.enabled && !this.config.allow_unknown_host){
        this.appssl.use(check_host(defOptions));
      }
      if (defOptions.enable_hsts == true){
        //ALL COPYRIGHT OF EXPRESS-FORCE-HTTPS PACKAGE.
        //THIS METHOD WAS IMPROVED BECAUSE IT'S ORIGINALLY VULNERABLE.
        this.app.use(function(req,res,next) {
          let schema = (req.headers['x-forwarded-proto'] || '').toLowerCase();
          if (String(req.headers.host).indexOf('localhost')<0 && schema!=='https') {
            return res.redirect('https://' + String(req.headers.host) + String(req.url));
          } else {
            return next();
          }
        });
        this.appssl.use(ProxyServer(defOptions));
      }
      else{
        this.app.use(ProxyServer(defOptions));
      }
    }
  }
  return proxy;
}

module.exports = createProxy;
